### PR TITLE
Python tests

### PR DESCRIPTION
**Pull request by manner82**
Some characteristic tests written to the python plugin support.
Also contains a fix for IO plugin validate() call, which I have found during the testing.
And adds a little example script on how to generate code coverage.

### CODE REVIEW

**Comment by millert** [comment]
When running "make check" in plugins/python I get the following libtool warning:

*** Warning: Linking the executable check_python_examples against the loadable module
*** python_plugin.so is not portable!

**Comment by millert** [comment]
On OpenBSD at least I see the following:
```
libtool: link: gcc -std=gnu99 -o .libs/check_python_examples check_python_examples.o iohelpers.o testhelpers.o -Wl,-z -Wl,relro -fstack-protector-strong  -L/home/millert/tmp/python_tests/lib/util/.libs -L/usr/local/lib -L../../lib/util/.libs -L./.libs -lpython_plugin -lpython3.7m -lintl -liconv -lpthread -lutil -lm -lsudo_util -Wl,-rpath,/usr/local/libexec/sudo -Wl,-rpath,/usr/local/lib
ld: error: unable to find library -lpython_plugin
```
I think it would be best to use sudo_dso_load() to load the plugin instead of trying to link with it directly.  The sudoers check_symbols test (sudoers/regress/check_symbols/check_symbols.c) does this and is passed the plugin path in argv[1] from the Makefile.

File: plugins/python/regress/testdata/check_example_io_plugin_command_log.stored [hunk 1]
```diff
@@ -0,0 +1,16 @@
+ -- Plugin STARTED --
+EXEC id --help
+EXEC info [
+    "command=/bin/id",
+    "runas_uid=0"
+]
+STD IN some standard input
+STD OUT some standard output
+STD ERR some standard error
+SUSPEND SIGTSTP (20)
```
Comment (millert) [inline plugins/python/regress/testdata/check_example_io_plugin_command_log.stored]: signal numbers are not portable so this test will fail on non-Linux systems

File: plugins/python/regress/testdata/check_example_conversation_plugin_reason_log_with_suspend.stdout [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Please provide your reason for executing ('/bin/whoami',)
+conversation suspend: signal SIGTSTP (20)
```
Comment (millert) [inline plugins/python/regress/testdata/check_example_conversation_plugin_reason_log_with_suspend.stdout]: signal numbers are not portable so this test will fail on non-Linux

File: plugins/python/regress/testdata/check_example_io_plugin_command_log.stored (hunk 1 above)
Comment (manner82) [reply plugins/python/regress/testdata/check_example_io_plugin_command_log.stored]: fixed

**Comment by millert** [comment]
> I think it would be best to use sudo_dso_load() to load the plugin instead of trying to link with it directly. The sudoers check_symbols test (sudoers/regress/check_symbols/check_symbols.c) does this and is passed the plugin path in argv[1] from the Makefile.

Or just link with $(OBJS) and $(LIBS) instead.

File: plugins/python/Makefile.in [hunk 3]
```diff
@@ -44,8 +44,10 @@ INSTALL_BACKUP = @INSTALL_BACKUP@
 LT_LIBS = $(top_builddir)/lib/util/libsudo_util.la
 LIBS = $(LT_LIBS)
 
+LIBPYTHONPLUGIN = python_plugin.la
+
 # C preprocessor flags
-CPPFLAGS = -I$(incdir) -I$(top_builddir) -I$(top_srcdir) @CPPFLAGS@ @PYTHON_INCLUDE@
+CPPFLAGS = -I$(incdir) -I$(top_builddir) -I$(top_srcdir) -DSRC_DIR=\"$(abspath $(srcdir))\" @CPPFLAGS@ @PYTHON_INCLUDE@
```
Comment (millert) [inline plugins/python/Makefile.in]: Using $(abspath) creates a dependency on GNU make.  Currently sudo will build with any version of make and I'd like it to stay that way.  You can @abs_srcdir@ from configure instead by defining abs_srcdir in Makefile.in near where srcdir is defined.  E.g.

```
srcdir = @srcdir@
devdir = @devdir@
top_builddir = @top_builddir@
top_srcdir = @top_srcdir@
abs_srcdir = @abs_srcdir@

...

CPPFLAGS = -I$(incdir) -I$(top_builddir) -I$(top_srcdir) -DSRC_DIR=\"$(abs_srcdir)\" @CPPFLAGS@ @PYTHON_INCLUDE@
```

```

Comment (millert) [reply plugins/python/Makefile.in]: This can be changed as part of a larger diff to use the absolute versions of src and build dirs where needed.

**Comment by manner82** [comment]
Replacing "abspath" with "abs_srcdir" seems to work fine. Also rebased the patches.

**Pull request merged**